azure/services/async/helper.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package async
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha4"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/converters"
+	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
+)
+
+// ProcessOngoingOperation is a helper function that will process an ongoing operation to check if it is done.
+// If it is not done, it will return a transient error.
+func ProcessOngoingOperation(ctx context.Context, scope FutureScope, client FutureHandler, resourceName string, serviceName string) error {
+	future := scope.GetLongRunningOperationState(resourceName, serviceName)
+	if future == nil {
+		scope.V(2).Info("no long running operation found", "service", serviceName, "resource", resourceName)
+		return nil
+	}
+	sdkFuture, err := converters.FutureToSDK(*future)
+	if err != nil {
+		// Reset the future data to avoid getting stuck in a bad loop.
+		scope.DeleteLongRunningOperationState(resourceName, serviceName)
+		return errors.Wrap(err, "could not decode future data, resetting long-running operation state")
+	}
+	done, err := client.IsDone(ctx, sdkFuture)
+	if err != nil {
+		return errors.Wrap(err, "failed checking if the operation was complete")
+	}
+
+	if !done {
+		// Operation is still in progress, update conditions and requeue.
+		scope.V(2).Info("long running operation is still ongoing", "service", serviceName, "resource", resourceName)
+		return azure.WithTransientError(azure.NewOperationNotDoneError(future), reconciler.DefaultReconcilerRequeue)
```

Comment:
We may want to consider using the `max(reconciler.DefaultReconcilerRequeue, done.AzureRetryAfter)`. I'm not sure if that exists on `done`, but the `RETRY-AFTER` header should be respected.

Furthermore, we might be able to encourage Azure folks to make `RETRY-AFTER` more useful to inform us how long they think it will take to complete the operation.

Reply:
I will look into this

Reply:
added handling of retry after